x-pack/plugins/security_solution/server/endpoint/lib/policy/license_watch.ts

```diff
@@ -0,0 +1,101 @@
+import { PackagePolicyServiceInterface } from '../../../../../fleet/server';
+import { ILicense } from '../../../../../licensing/common/types';
+import {
+  isEndpointPolicyValidForLicense,
+  unsetPolicyFeaturesAboveLicenseLevel,
+} from '../../../../common/license/policy_config';
+import { licenseService } from '../../../lib/license/license';
+
+export class PolicyWatcher {
+  private logger: Logger;
+  private soClient: SavedObjectsClientContract;
+  private policyService: PackagePolicyServiceInterface;
+  private subscription: Subscription | undefined;
+  constructor(
+    policyService: PackagePolicyServiceInterface,
+    soStart: SavedObjectsServiceStart,
+    logger: Logger
+  ) {
+    this.policyService = policyService;
+    this.soClient = this.makeInternalSOClient(soStart);
+    this.logger = logger;
+  }
+
+  /**
+   * The policy watcher is not called as part of a HTTP request chain, where the
+   * request-scoped SOClient could be passed down. It is called via license observable
+   * changes. We are acting as the 'system' in response to license changes, so we are
+   * intentionally using the system user here. Be very aware of what you are using this
+   * client to do
+   */
+  private makeInternalSOClient(soStart: SavedObjectsServiceStart): SavedObjectsClientContract {
+    const fakeRequest = ({
+      headers: {},
+      getBasePath: () => '',
+      path: '/',
+      route: { settings: {} },
+      url: { href: {} },
+      raw: { req: { url: '/' } },
+    } as unknown) as KibanaRequest;
+    return soStart.getScopedClient(fakeRequest, { excludedWrappers: ['security'] });
+  }
+
+  public start() {
+    this.subscription = licenseService.getLicenseInformation$()?.subscribe(this.watch.bind(this));
+  }
+
+  public stop() {
+    if (this.subscription) {
+      this.subscription.unsubscribe();
+    }
+  }
+
+  public async watch(license: ILicense) {
+    let page = 1;
+    let response: {
+      items: PackagePolicy[];
+      total: number;
+      page: number;
+      perPage: number;
+    };
+    do {
+      try {
+        response = await this.policyService.list(this.soClient, {
+          page: page++,
+          perPage: 100,
+          kuery: `${PACKAGE_POLICY_SAVED_OBJECT_TYPE}.package.name: endpoint`,
+        });
+      } catch (e) {
+        this.logger.warn(
+          `Unable to verify endpoint policies in line with license change: failed to fetch package policies: ${e.message}`
+        );
+        return;
+      }
+      response.items.forEach((policy) => {
+        const policyConfig = policy.inputs[0].config?.policy.value;
+        if (!isEndpointPolicyValidForLicense(policyConfig, licenseService)) {
+          policy.inputs[0].config!.policy.value = unsetPolicyFeaturesAboveLicenseLevel(
+            policyConfig,
+            licenseService
+          );
+          this.policyService.update(this.soClient, policy.id, policy);
```

Comment:
Hmm.... Should you be throttling here? You could be generating several calls in parallel/rapid succession. 

Also - you need to account for failures here for two reasons:

1. to ensure that the remainder of the loop succeeds. You also don't want to cause the outer look (`do{} while()`) to throw, I don't think (not sure what that will do to the observable calling logic).
2. to account for update conflicts, which may be the case if an update is done right after you retrieve the list of items or by other Kibana instance that may also be doing this same thing. I think we said maybe you do 1 attempt after a failure and then just log it if it still fails.

Reply:
Do we have an accepted practice for how to throttle? This is certainly could be firing off a lot of update calls.

If I wrap `update` in a try/catch, do I now need to `await update` to make sure it's actually caught? If I'm now awaiting, do I still need to throttle since they are no longer in parallel?

Reply:
Re:
> Do we have an accepted practice for how to throttle?

I did a little of this in Fleet using Bluebird - see here https://github.com/elastic/kibana/blob/95861a0fb0198bbfbd6f82b380caf86e911d97cd/x-pack/plugins/fleet/server/routes/agent_policy/handlers.ts#L55-L65

Being that you are paging through, you might able to use a variation of that approach (chain each page to do concurrent updates, then `await` at the end).